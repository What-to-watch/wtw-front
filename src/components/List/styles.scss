$fontColor: #FFFFFF;
$backgorund: #131414;
$hover: #424343;
$selected: #183E38;
$lineColor: rgba(255, 255, 255, 0.4);
$radius: 8px;

.list {
    display: flex;
    flex-direction: column;
    justify-content: flex-start;
    align-items: flex-start;
    box-shadow: 0px 4px 8px rgba(0, 0, 0, 0.24);
    border-radius: $radius;
    min-width: 220px;
    list-style: none;

    &__item:nth-of-type(1) {
        border-top-left-radius: $radius;
        border-top-right-radius: $radius;
    }

    &__item:nth-last-of-type(1) {
        border-bottom-left-radius: $radius;
        border-bottom-right-radius: $radius;
        border-bottom: none;
    }

    &__item {
        width: 100%;
        display: flex;
        justify-content: flex-start;
        align-items: center;
        flex-direction: row;
        border-bottom: 1px solid $lineColor;
        background-color: $backgorund;
        font-family: Roboto;
        font-weight: normal;
        font-size: 18px;
        line-height: 21px;
        color: $fontColor;
        padding: 16px 24px;

        &:hover {
            background-color: $hover;
        }

        &:active {
            background-color: $selected;
        }
    }
}