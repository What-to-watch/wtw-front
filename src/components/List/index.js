import React from 'react';
import PropTypes from 'prop-types';

import "./styles.scss";

const mapToItems = (data,props) => 
    data
    .map(x => typeof x === "string" ? { label: x, value: x } : x )
    .map(({label, value, ...meta},index) => <Item 
        index={index} 
        key={index} 
        label={label}
        value={value}
        {...meta}
        {...props} 
    />
    )


const mapToHighlights = (str,keyword) => {
    const keywordRegExp = new RegExp(keyword,"gi")
    if( keyword && keyword.trim() && keywordRegExp.test(str) ){
        return <span>{str.replace(keywordRegExp, str => `|${str}|`)
                .split("|")
                .map((str,idx) => {
                    if(str.toLowerCase() === keyword.trim().toLowerCase()){
                        return <strong key={`highlight-${idx}`}>{str}</strong>
                    } else {
                        return str
                    }
                })}</span>
    } else {
        return str
    }
}

const overrideItemProps = (children,props) => React.Children.map(children,(child) => {
    return React.cloneElement(child,{
        ...child.props,
        ...props,
    })
})

export const Item = (props) => {
    const { label, value, keyword, children, onClick = x => x } = props;
    const text = label || children || value;

    const handleClick = () => {
        onClick(value,props)
    }

    return <li className="list__item" onClick={handleClick}>
        {mapToHighlights(text,keyword)}
    </li>
}

const List = (props) => {
    const { 
        content, 
        keyword, 
        children, 
        onItemClick = x => x
    } = props;

    const itemProps = {
        keyword,
        onClick: onItemClick
    }

    return <ul className="list" >
        {content ? mapToItems(content,itemProps) : overrideItemProps(children,itemProps)}
    </ul>
}

List.Item = Item;

Item.propTypes = {
    value: PropTypes.node,
    keyword: PropTypes.string,
    children: PropTypes.node,
    onClick: PropTypes.func,
}

List.propTypes = {
    keyword: PropTypes.string,
    content: PropTypes.arrayOf(PropTypes.oneOfType([
        PropTypes.shape({
            label: PropTypes.node.isRequired,
            value: PropTypes.any.isRequired
        }),
        PropTypes.node
    ])),
    children: (props) => {
        if( props.children ){
            if(props.content){
                throw new Error("List should have either content or children. Not both.")
            }
            if(props.children.some(p => p.type !== Item)){
                const others = props.children.filter(p => p.type !== Item).map(x => x.type)
                throw new Error(`List may only contain Item children. Found: ${others}`)
            }
        }
    },
    onItemClick: PropTypes.func,
}

export default List;