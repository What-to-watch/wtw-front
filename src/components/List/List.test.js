import React from 'react';
import { render, fireEvent, screen } from '@testing-library/react'
import renderer from 'react-test-renderer';
import List, { Item } from './';

describe("List rendering", () => {
    it("renders correctly", () => {
        const content = [
            "Comedy", "Drama", "Documentary"
        ]
        const comp = render(<List content={content}/>)
        expect(comp.asFragment()).toMatchSnapshot();
    })

    it("content prop rendering", () => {
        const content = [
            "Comedy", "Drama", "Documentary"
        ]
        const comp = render(<List content={content}/>)

        expect(comp.queryAllByRole("list")).toHaveLength(1);
        expect(comp.queryAllByRole("listitem")).toHaveLength(content.length);
    })

    it("children rendering", () => {
        const content = [
            "Comedy", "Drama", "Documentary"
        ]
        const comp = render(<List>
            <Item>Comedy</Item>
            <Item>Drama</Item>
            <Item>Documentary</Item>
        </List>)

        expect(comp.queryAllByRole("list")).toHaveLength(1);
        expect(comp.queryAllByRole("listitem")).toHaveLength(content.length);
    })
})

describe("List behavior", () => {
    it("should highligh given a search prop (case insensitive)", () => {
        const content = [
            "Coca cola", "Comedy", "Pintuco"
        ]
        const keyword = "co";
        const comp = renderer.create(<List content={content} keyword={keyword}/>)
        expect(comp.root.findAllByType("strong")).toHaveLength(4)
    })

    it("should receive value and props on click", () => {
        const content = [
            { label: "Coca cola", value: "Something else" }
        ]
        const spy = (() => {
            let _spy = (...args) => {
                _spy.calls.push({ args })
                _spy.called = true;
                _spy.callCount = _spy.callCount+1;
            }
            _spy.calls = [];
            _spy.called = false;
            _spy.callCount = 0;
            Object.defineProperty(_spy,"firstCall",{
                get: () => _spy.calls[0]
            })
            return _spy
        })()
        render(<List content={content} onItemClick={spy} />)
        fireEvent.click(screen.getByText('Coca cola'))

        expect(spy.called).toBeTruthy();
        expect(spy.callCount).toBe(1);
        expect(spy.firstCall.args[0]).toBe(content[0].value)
    })
})