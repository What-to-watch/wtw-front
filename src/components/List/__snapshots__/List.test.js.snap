// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`List rendering renders correctly 1`] = `
<DocumentFragment>
  <ul
    class="list"
  >
    <li
      class="list__item"
    >
      Comedy
    </li>
    <li
      class="list__item"
    >
      Drama
    </li>
    <li
      class="list__item"
    >
      Documentary
    </li>
  </ul>
</DocumentFragment>
`;
